src/vault.py

```diff
@@ -0,0 +1,282 @@
+        self.client.write(
+            "{}/roles/{}".format(CHARM_PKI_MP, CHARM_PKI_ROLE), server_flag=True, **kwargs
+        )
+        # Configure a role for using this PKI to issue client-only certs
+        self.client.write(
+            "{}/roles/{}".format(CHARM_PKI_MP, CHARM_PKI_ROLE_CLIENT),
+            server_flag=False,  # client certs cannot be used as server certs
+            **kwargs,
+        )
+
+    def sign_csr(self, csr: str, ttl="87599h"):
+        """Signs CSR.
+
+        :param csr: Certificate Signing Request
+        :param ttl: Certificate Time To Live
+        """
+        logger.info("Vault - Signing CSR")
+        return self.client.write(
+            "{}/root/sign-intermediate".format(CHARM_PKI_MP), csr=csr, format="pem_bundle", ttl=ttl
+        )["data"]["certificate"]
+
+    def issue_certificate(self, common_name: str, sans: str, cert_type: str) -> dict:
+        """Issues a key and certificate to a requesting charm.
+
+        The certificate_data should contain "certificate_name",
+        "common_name", and "sans"
+        """
+        logger.info("Vault - Issuing certificate")
+        if cert_type == "server":
+            role = CHARM_PKI_ROLE
+        elif cert_type == "client":
+            role = CHARM_PKI_ROLE_CLIENT
+        else:
+            raise RuntimeError("Unsupported cert_type: " "{}".format(cert_type))
+        config = {
+            "common_name": common_name,
+        }
+        if sans:
+            sans = json.loads(sans)
+            ip_sans, alt_names = self._sort_sans(sans)  # type: ignore[arg-type]
+            if ip_sans:
+                config["ip_sans"] = ",".join(ip_sans)
+            if alt_names:
+                config["alt_names"] = ",".join(alt_names)
+        try:
+            logging.info("About to create a certificate with {}".format(config))
+            response = self.client.write("{}/issue/{}".format(CHARM_PKI_MP, role), **config)
+            if not response["data"]:
+                raise RuntimeError(response.get("warnings", "unknown error"))
+        except hvac.exceptions.InvalidRequest as e:
+            raise RuntimeError(str(e)) from e
+        logging.info(f"new cert data: {response['data']}")
+        return response["data"]
+
+    def get_token(self, name: str, cidr: List[str]) -> dict:
+        """Returns token to communicate with Vault.
+
+        :param name: Role name
+        :param cidr: List of CIDR's to be allowed to communicate with Vault.
+        """
+        logger.info("Vault - Getting token")
+        return self.client.write(
+            "auth/approle/role/{}/secret-id".format(name), wrap_ttl="1h", cidr_list=cidr
+        )
+
+    def create_role(self, approle_name: str, policy_name: str, cidr: List[str]) -> None:
+        """Creates role in Vault.
+
+        :param approle_name: Specifies the name of the role to create.
+        :param policy_name: Policy OID.
+        :param cidr: List of CIDR's.
+        """
```

Comment:
Docstrings should follow google format - applies elsewhere too :)

Reply:
Agreed, fixed.